Modify ByteBufferReadableStreamChannel to take in a list of ByteBuffer

moontails opened the pull request:
Modify ByteBufferReadableStreamChannel to take in a list of ByteBuffer (#394).

Extending ByteBufferReadableStreamChannel to take in a `List` of `ByteBuffer`. 
- Updating existing constructor to call into the new constructor.
- Updating class member `ByteBuffer buffer` to `List<ByteBuffer> buffers`
- Updating `readInto()` to process a `List` of `ByteBuffer`

TODO 
- add/update tests.

vgkholla commented on ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java:
```diff
@@ -30,16 +36,34 @@
+  private final ReentrantLock contentLock = new ReentrantLock();
+  private final List<ByteBuffer> buffers;
+
+  private int size;
```
can be `final`

vgkholla commented on ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java:
```diff
@@ -30,16 +36,34 @@
+   */
+
+  public ByteBufferReadableStreamChannel(List<ByteBuffer> bufferList) {
+    if (bufferList != null) {
```
`bufferList` should not be `null` - you should throw a `IllegalArgumentException`

vgkholla commented on ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java:
```diff
@@ -30,16 +36,34 @@
+      buffers = new ArrayList<>();
+    }
+    // null to signify end of the list
+    buffers.add(null);
```
since you are not doing `addContent()` for now, why do you need this end marker? Should be easy to determine the "last"content right?

moontails replied:
Makes sense, updated code.

vgkholla commented on ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java:
```diff
@@ -49,21 +73,42 @@ public long getSize() {
-      if (callback != null) {
-        callback.onCompletion(0L, closedChannelException);
+    ReadIntoCallbackWrapper callbackWrapper = new ReadIntoCallbackWrapper(callback);
+    contentLock.lock();
```
don't need a lock if you are not doing `addContent()`. An `AtomicBoolean` and the `compareAndSet` method should do.

vgkholla replied:
`channelEmptied` (which exists already) should be enough for this

vgkholla commented on ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java:
```diff
@@ -76,4 +121,92 @@ public void close()
     channelOpen.set(false);
   }
+
+  private int computeSize(List<ByteBuffer> bufferList) {
```
add documentation

vgkholla commented on ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java:
```diff
@@ -76,4 +121,92 @@ public void close()
   }
+
+  private int computeSize(List<ByteBuffer> bufferList) {
+    int size = 0;
```
avoid having local variable names that have the same name as class variables

moontails commented on ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java:
```diff
@@ -49,21 +73,42 @@ public long getSize() {
+  private void writeContent(AsyncWritableChannel writeChannel, ReadIntoCallbackWrapper callbackWrapper,
+                            ByteBuffer content) {
+    ContentWriteCallback writeCallback;
+    if (content == null) {
```
Can you confirm if the null check is needed here? Or is it safe to assume that there will be no null fields in the list of bytebuffers?

vgkholla replied:
there cannot be a `null` in the list of `ByteBuffer`. If there was one, `buffer.array().length` in `computeSize()` would have thrown a `NullPointerException` and the object would not even have been constructed.

moontails replied:
there is a actually a null check in `computeSize()`. So my question is does it need to be handled or should it have thrown a NullPointerException and the object should not be constructed?

vgkholla replied:
Should have thrown `NullPointerException`. It doesn't make logical sense to have null `ByteBuffer` instances inbetween.

vgkholla commented on ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java:
```diff
@@ -76,4 +121,92 @@ public void close()
+    int size = 0;
+    for (ByteBuffer buffer : bufferList) {
+      if (buffer != null ) {
+        size += buffer.array().length;
```
use `buffer.remaining()`.

moontails commented:
Addressed the following

- Marked `size` as final
- Added documentaion to `computeSize`, using `buffer.remaning()`
- Removed lock mechanism
- Removed explicit null checks in `writeContent` and `computeSize`
- Removed appending null as last block

vgkholla commented:
@moontails will review this today or tomorrow. Sorry for the delay

moontails commented:
@vgkholla no problem and thank you!

I have also added a testcase for the corner cases involving the creation of `ByteBufferReadableStreamChannel`. 

Since we are effectively calling the second constructor from the first, I was thinking the existing tests would do as they are now run on a list of `ByteBuffer`. What do you think?

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferReadableStreamChannelTest.java:
```diff
@@ -177,10 +180,75 @@ public void idempotentOperationsTest()
+   * 4. Create with a list of {@link ByteBuffer} containing a null {@link ByteBuffer}.
+   */
+  @Test
+  public void cornerCasesTest() {
```
better called `constructionErrorCasesTest()` since these are not "corner" cases.

moontails replied:
renamed the testcase

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferReadableStreamChannelTest.java:
```diff
@@ -177,10 +180,75 @@ public void idempotentOperationsTest()
+    ByteBufferReadableStreamChannel readableStreamChannel;
+    try {
+      ByteBuffer content = null;
+      readableStreamChannel = new ByteBufferReadableStreamChannel(content);
```
whenever an operation _has_ to fail, the last line of the `try` block has to be `fail("Operation should have failed")`. Otherwise, notice that this test will succeed even if no exception was thrown.

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferReadableStreamChannelTest.java:
```diff
@@ -177,10 +180,75 @@ public void idempotentOperationsTest()
+    try {
+      ByteBuffer content = null;
+      readableStreamChannel = new ByteBufferReadableStreamChannel(content);
+    } catch (RuntimeException e) {
```
why not catch `NullPointerException`? (or `IllegalArgumentException`)

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferReadableStreamChannelTest.java:
```diff
@@ -177,10 +180,75 @@ public void idempotentOperationsTest()
+    // null byteBuffer
+    ByteBufferReadableStreamChannel readableStreamChannel;
+    try {
+      ByteBuffer content = null;
```
include _only_ the statement that is expected to fail in the `try` and nothing else.

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferReadableStreamChannelTest.java:
```diff
@@ -177,10 +180,75 @@ public void idempotentOperationsTest()
+
+    // null bufferList
+    List<ByteBuffer> bufferList;
+    try {
```
there is a lot of repeated code in the `try..catch` blocks here. Please try not to repeat code if possible. For example, this can possibly be made into a function that is something like 
```
 private void constructionFailureTest(List<ByteBuffer> bufferList, Class expectedExceptionType) {
    try {
      new ByteBufferReadableStreamChannel(bufferList);
      fail("Operations should have failed");
    } catch (RuntimeException e) {
      assertEquals("Unexpected RuntimeException", expectedExceptionType, e.getClass());
    }
  }
```
You don't need the root causing either I think.

moontails replied:
I like this idea, as you say removes some of the repeated code. Will update the test.

moontails replied:
Have updated the tests

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferReadableStreamChannelTest.java:
```diff
@@ -177,10 +180,75 @@ public void idempotentOperationsTest()
+   * @param contents the variable {@code ByteBuffer} argument.
+   * @return list of {@code ByteBuffer}.
+   */
+  private List<ByteBuffer> fillBufferList(Object... contents) {
```
wouldn't `Arrays.asList()` or `new ArrayList(Arrays.asList())` serve the same purpose as this function? Even if they don't allow `null` values, it seems like overkill to have a function for a single use case.

vgkholla commented:
@moontails The existing tests would run on the new constructor but they wouldn't test list specific use cases - for e.g., only the last `ByteBuffer` in a list should go in with `isLast` as `true`. Since you never have more than one item in the list in the current test cases, there is no opportunity to test the code paths that occur when `isLast` is `false`.

moontails commented on ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java:
```diff
@@ -30,16 +36,25 @@
+    if (bufferList == null || bufferList.isEmpty()) {
+      throw new IllegalArgumentException("Input List of ByteBuffer cannot be null/empty for ByteBufferReadableStreamChannel");
+    }
+    buffers = new ArrayList<>(bufferList);
```
If at all the`ByteBufferReadableStreamChannel` is constructed using an immutable list, then it would break at`bufferIterator.remove();` in `readInto`.

vgkholla replied:
Ok - why do you need the `remove()`? Correct me if I'm wrong, but it doesn't looks necessary. `isLast` can be determined in other ways.

moontails replied:
`remove()` was not added for the sake of `isLast`. 

In the #394 you suggested taking a look at `MockRestRequest` ([link](https://github.com/linkedin/ambry/blob/master/ambry-api/src/test/java/com.github.ambry/rest/MockRestRequest.java#L194)). Looking at that code, i thought it was imperative to remove the buffer from the list, once it has been processed.

vgkholla replied:
Yes, but notice that in `MockRestRequest`, buffers can be added after construction (which cannot be done here) - so assuming the list is mutable is fair in `MockRestRequest`.

Aside from that, it might not even be important to `remove()` in `MockRestRequest`. It was probably done for a use-case that may no longer be true. When you encounter situations like this, it is best to think about whether it makes sense for this use case (and for the future). Looking at it from that point of view, even if this class supported adding buffers in the future, `remove()` may not be required. If it is, then assuming a mutable list would be a fair thing to do.

moontails replied:
Coolio. Thank you for the clarification!

moontails commented on ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java:
```diff
@@ -30,16 +36,25 @@
-  public ByteBufferReadableStreamChannel(ByteBuffer buffer) {
-    this.buffer = buffer;
-    size = buffer.remaining();
+  public ByteBufferReadableStreamChannel(ByteBuffer buffer) { this(Arrays.asList(buffer)); }
```
thank for spotting this, i have updated the code

moontails commented:
@vgkholla I apologize for the delay in making the requested changes. Have split the original test into 

- `ByteBufferReadableStreamChannelTest` that contains only the corner cases tests
- `ByteBufferReadableStreamChannelParameterizedTest` that contains the other test cases to test against different ways of creating a ByteBufferReadableStreamChannel

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferReadableStreamChannelParameterizedTest.java:
```diff
@@ -0,0 +1,280 @@
+  private ByteBuffer contentWrapper;
+  private ByteBufferReadableStreamChannel byteBufferReadableStreamChannel;
+
+  public ByteBufferReadableStreamChannelParameterizedTest(List<ByteBuffer> bufferList, ByteBuffer content1,
```
instead of passing the verifying buffers as parameters, you could just create a copy of what is in the list, no?

moontails replied:
Makes sense, will update

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferReadableStreamChannelTest.java:
```diff
@@ -38,93 +30,6 @@
 public class ByteBufferReadableStreamChannelTest {
```
i'd suggest naming this as `ByteBufferReadableStreamChannelCommonTest` and naming the other one as `ByteBufferReadableStreamChannelTest`

moontails replied:
Will rename the files

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferReadableStreamChannelTest.java:
```diff
@@ -156,118 +61,60 @@ public void readAndWriteCornerCasesTest() throws Exception {
+    try {
+      new ByteBufferReadableStreamChannel((ByteBuffer) null);
+      fail("Construction of ByteBufferReadableStreamChannel should have failed");
+    } catch (RuntimeException e) {
```
can just catch `NullPointerException` here.

moontails replied:
This is to ascertain the right exception is thrown.

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferReadableStreamChannelTest.java:
```diff
@@ -156,118 +61,60 @@ public void readAndWriteCornerCasesTest() throws Exception {
-      throw new TimeoutException("Waiting too long for callback to arrive");
-    }
+    // bufferList containing a null byteBuffer
+    List<ByteBuffer> bufferList = Arrays.asList(ByteBuffer.wrap(fillRandomBytes(new byte[32])), null);
```
can you use `Utils.getRandomBytes()` and get rid of the `fillRandomBytes()` function?

moontails replied:
Replaced with `TestUtils.getRandomBytes()`

pnarayanan commented:
Hey - I wish to take a quick look, could you hold off merging for now?
On May 2, 2017 10:05 PM, "Gopalakrishna Holla" <notifications@github.com>
wrote:

> *@vgkholla* approved this pull request.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/linkedin/ambry/pull/492#pullrequestreview-35940248>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYzQ17o8atNHbuDXVdq4WawOsLqlEeuks5r2ArtgaJpZM4KyKMD>
> .
>

moontails commented:
Hi @pnarayanan did you get a chance to review?

pnarayanan commented on ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java:
```diff
@@ -30,16 +33,28 @@
+   * Constructs a {@link ReadableStreamChannel} whose read operations return data from the provided {@code bufferList}.
+   * @param bufferList list of {@link ByteBuffer} that is used to retrieve data from on invocation of read operations.
+   */
+  public ByteBufferReadableStreamChannel(List<ByteBuffer> bufferList) {
```
@vgkholla @moontails what did you decide about also letting this class take in buffers after construction?

vgkholla replied:
@pnarayanan Seemed to be a larger change with no use case right now. Once this is merged, we can consider doing that as an extension

pnarayanan commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferReadableStreamChannelTest.java:
```diff
@@ -28,33 +31,59 @@
+  public static List<Object[]> data() {
+    // @formatter:off
+    return Arrays.asList(new Object[][] {
+      {Arrays.asList(ByteBuffer.wrap(TestUtils.getRandomBytes(BUFFER_SIZE)))},
```
You could keep this simple and just pass `{true, false}` as you seem to have to do a check in the constructor anyway.

Also, do you think there a value in randomizing (with a reasonable max) the number of buffers in the list, rather than hardcoding it to 2?

pnarayanan commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferReadableStreamChannelCommonTest.java:
```diff
@@ -0,0 +1,111 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.commons;
```
Are these tests logically separate from the tests in the other file? If this is just due to the parametrization in the other file, I think you can work around that and combine these into one file.

pnarayanan commented:
@moontails this looks mostly good to me, once you address these last comments this can be merged.

Please also rebase this PR with the master branch.

pnarayanan commented:
Closing this due to inactivity. If we decide to take this change up again, a new PR may be opened using this one as a reference.
